kernel/printf.c

```diff
@@ -0,0 +1,249 @@
+#include "printf.h"
+
+#include "com.h"
+#include "util.h"
+
+#include <stdarg.h>
+#include <stdint.h>
+
+#define emit(c) do { com_putch(c); ++ret; } while(0)
+
+#define CONV_LEFT_JUST 0x01
+#define CONV_USE_PLUS  0x02
+#define CONV_USE_SPACE 0x04
+#define CONV_ZERO_PAD  0x08
+
+enum arg_size {
+  SIZE_SHORT,
+  SIZE_DEFAULT,
+  SIZE_LONG,
+};
+
+int printf (const char *fmt, ...) {
```

Comment:
The block nesting in this code is hard to read and reason about (it goes multiple pages).

Is there any possibility to factor some of this up into functions?